File: app/src/main/java/com/mandarin/bcu/androidutil/lineup/adapters/LUUnitList.java
package com.mandarin.bcu.androidutil.lineup.adapters;

import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.mandarin.bcu.R;
import com.mandarin.bcu.androidutil.FilterEntity;
import com.mandarin.bcu.androidutil.StaticStore;
import com.mandarin.bcu.androidutil.lineup.LineUpView;
import com.mandarin.bcu.androidutil.unit.adapters.UnitListAdapter;

import java.util.ArrayList;

import common.battle.BasisSet;
import common.util.unit.Form;
import common.util.unit.Unit;

public class LUUnitList extends Fragment {
    private LineUpView line;
    private Handler handler = new Handler();
    private Runnable runnable;

    public static LUUnitList newInstance(String[] names, LineUpView line) {
        LUUnitList ulist = new LUUnitList();
        Bundle bundle = new Bundle();
        bundle.putStringArray("Names", names);
        ulist.setArguments(bundle);
        ulist.setLineUp(line);

        return ulist;
    }

    private boolean destroyed = false;
    private ArrayList<Integer> numbers = new ArrayList<>();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup group, @Nullable Bundle bundle) {
        View view = inflater.inflate(R.layout.lineup_unit_list, group, false);

        if (line == null) {
            if (getActivity() != null)
                line = getActivity().findViewById(R.id.lineupView);
        }

        if (getArguments() == null) return view;

        FilterEntity entity = new FilterEntity(StaticStore.unitnumber);
        numbers = entity.setFilter();
        ArrayList<String> names = new ArrayList<>();

        for (int i : numbers) {
            names.add(StaticStore.LUnames[i]);
        }

        UnitListAdapter adapter = new UnitListAdapter(getActivity(), names.toArray(new String[0]), numbers);

        ListView ulist = view.findViewById(R.id.lineupunitlist);

        ulist.setAdapter(adapter);

        runnable = new Runnable() {
            @Override
            public void run() {
                if (StaticStore.updateList) {
                    FilterEntity entity = new FilterEntity(StaticStore.unitnumber);
                    numbers.clear();
                    numbers = entity.setFilter();
                    ArrayList<String> names = new ArrayList<>();

                    for (int i : numbers) {
                        names.add(StaticStore.LUnames[i]);
                    }

                    UnitListAdapter adapter = new UnitListAdapter(getActivity(), names.toArray(new String[0]), numbers);

                    ulist.setAdapter(adapter);

                    StaticStore.updateList = false;
                }

                if (!destroyed)
                    handler.postDelayed(this, 50);
            }
        };

        handler.postDelayed(runnable, 50);

        ulist.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Form f = StaticStore.units.get(numbers.get(position)).forms[StaticStore.units.get(numbers.get(position)).forms.length - 1];

                if (alreadyExist(f)) return;

                int[] posit = StaticStore.getPossiblePosition(BasisSet.current.sele.lu.fs);

                if (posit[0] != 100)
                    BasisSet.current.sele.lu.fs[posit[0]][posit[1]] = f;
                else
                    line.repform = f;

                line.UpdateLineUp();
                line.toFormArray();
                line.invalidate();
            }
        });

        return view;
    }

    private boolean alreadyExist(Form form) {
        Unit u = form.unit;

        for (int i = 0; i < BasisSet.current.sele.lu.fs.length; i++) {
            for (int j = 0; j < BasisSet.current.sele.lu.fs[i].length; j++) {
                if (BasisSet.current.sele.lu.fs[i][j] == null) {
                    if (line.repform == null) return false;

                    return u.equals(line.repform.unit);
                }

                Unit u2 = BasisSet.current.sele.lu.fs[i][j].unit;

                if (u.equals(u2))
                    return true;
            }
        }

        return false;
    }

    @Override
    public void onDestroy() {
        destroyed = !destroyed;
        handler.removeCallbacks(runnable);
        super.onDestroy();
    }

    private void setLineUp(LineUpView line) {
        this.line = line;
    }
}

File: app/src/main/java/com/mandarin/bcu/androidutil/lineup/adapters/LUUnitList.kt
package com.mandarin.bcu.androidutil.lineup.adapters

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ListView
import androidx.core.view.children
import androidx.fragment.app.Fragment
import com.mandarin.bcu.R
import com.mandarin.bcu.androidutil.StaticStore
import com.mandarin.bcu.androidutil.filter.FilterEntity
import com.mandarin.bcu.androidutil.lineup.LineUpView
import common.battle.BasisSet
import common.pack.Identifier
import common.util.stage.Stage
import common.util.unit.Form
import common.util.unit.Unit

class LUUnitList : Fragment() {
    private lateinit var line: LineUpView
    private var stage: Stage? = null
    private var star = 0
    private var numbers = ArrayList<Identifier<Unit>>()

    override fun onCreateView(inflater: LayoutInflater, group: ViewGroup?, bundle: Bundle?): View? {
        val view = inflater.inflate(R.layout.lineup_unit_list, group, false)

        numbers = FilterEntity.setLuFilter()

        val adapter = LUUnitListAdapter(requireActivity(), numbers, stage)

        val ulist = view.findViewById<ListView>(R.id.lineupunitlist)

        ulist.adapter = adapter

        ulist.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            if(position < 0 || position >= numbers.size)
                return@OnItemClickListener

            val u = Identifier.get(numbers[position]) ?: return@OnItemClickListener

            val f = u.forms[u.forms.size - 1]

            if (alreadyExist(f))
                return@OnItemClickListener

            val posit = StaticStore.getPossiblePosition(BasisSet.current().sele.lu.fs)

            if (posit[0] != 100)
                BasisSet.current().sele.lu.fs[posit[0]][posit[1]] = f
            else
                line.repform = f

            line.updateLineUp()
            line.syncLineUp()
            line.invalidate()

            sync()
        }

        if (this::line.isInitialized) {
            adapter.lineup = line
        }

        return view
    }

    fun update() {
        val v = view ?: return

        val unitList = v.findViewById<ListView>(R.id.lineupunitlist)

        numbers.clear()
        numbers = FilterEntity.setLuFilter()

        val adapter1 = LUUnitListAdapter(requireActivity(), numbers, stage)

        unitList.adapter = adapter1
    }

    fun sync() {
        val v = view ?: return

        val unitList = v.findViewById<ListView>(R.id.lineupunitlist)
        val adapter = unitList.adapter ?: return

        if (adapter !is LUUnitListAdapter)
            return

        var index = unitList.firstVisiblePosition

        for (child in unitList.children) {
            val tag = child.tag

            if (tag is LUUnitListAdapter.ViewHolder) {
                tag.updateComponents(index++)
            }
        }
    }

    private fun alreadyExist(form: Form): Boolean {
        val u = form.unit

        for (i in BasisSet.current().sele.lu.fs.indices) {
            for (j in BasisSet.current().sele.lu.fs[i].indices) {

                if (BasisSet.current().sele.lu.fs[i][j] == null) {
                    return if (line.repform == null)
                        false
                    else
                        u == line.repform!!.unit

                }

                val u2 = BasisSet.current().sele.lu.fs[i][j].unit

                if (u == u2)
                    return true
            }
        }

        return false
    }

    fun setArguments(line: LineUpView, stage: Stage? = null, star: Int = 0) {
        this.line = line
        this.stage = stage
        this.star = star

        val v = view ?: return

        val unitList = v.findViewById<ListView>(R.id.lineupunitlist)
        val adapter = unitList.adapter ?: return

        if(adapter is LUUnitListAdapter) {
            adapter.lineup = line
        }
    }

    companion object {

        fun newInstance(line: LineUpView, stage: Stage? = null, star: Int = 0): LUUnitList {
            val ulist = LUUnitList()
            ulist.setArguments(line, stage, star)

            return ulist
        }
    }
}
